snippets/csharp/VS_Snippets_CLR/RegularExpressions.Examples.Email/cs/example4.cs

```diff
@@ -5,52 +5,48 @@
 
 public class RegexUtilities
 {
-   bool invalid = false;
+    public static bool IsValidEmail(string email)
+    {
+        if (string.IsNullOrEmpty(email))
+            return false;
 
-   public bool IsValidEmail(string strIn)
-   {
-       invalid = false;
-       if (String.IsNullOrEmpty(strIn))
-          return false;
+        // Use IdnMapping class to convert Unicode domain names.
+        try
+        {
+            // Examines the domain part of the email and normalizes it.
+            string DomainMapper(Match match)
+            {
+                // IdnMapping class with default property values.
+                var idn = new IdnMapping();
+
+                // Pull out and process domain name (throws ArgumentException on invalid)
+                var domainName = idn.GetAscii(match.Groups[2].Value);
 
-       // Use IdnMapping class to convert Unicode domain names.
-       try {
-          strIn = Regex.Replace(strIn, @"(@)(.+)$", this.DomainMapper,
-                                RegexOptions.None, TimeSpan.FromMilliseconds(200));
-       }
-       catch (RegexMatchTimeoutException) {
-         return false;
-       }
+                return match.Groups[1].Value + domainName;
+            }
 
-        if (invalid)
-           return false;
+            // Normalize the domain
+            email = Regex.Replace(email, @"(@)(.+)$", DomainMapper,
+                                  RegexOptions.None, TimeSpan.FromMilliseconds(200));
+        }
+        catch (Exception e) when (e is RegexMatchTimeoutException || e is ArgumentException)
```

Comment:
I'd probably use two different catch clauses here rather than an exception filter. You're only checking the type, not any properties of the exception, or any other state.